### PR TITLE
BUG Infinite loops in TempDatabase (fixes #8902)

### LINKED ISSUE #8902: Infinite loop in tests if temp database creation fails
If the temporary database creation fails for some reason, test executing locks up in an infinite loop. The root cause is this code, which is necessary for some edge-cases. I've come across this when a test table has a >64 character name.

```diff
diff --git a/src/ORM/Connect/TempDatabase.php b/src/ORM/Connect/TempDatabase.php
index a47c50078..e8f88e2b3 100644
--- a/src/ORM/Connect/TempDatabase.php
+++ b/src/ORM/Connect/TempDatabase.php
@@ -290,14 +290,14 @@ class TempDatabase
             return;
         }
 
-        try {
-            $this->rebuildTables($extraDataObjects);
-        } catch (DatabaseException $ex) {
-            // In case of error during build force a hard reset
-            // e.g. pgsql doesn't allow schema updates inside transactions
-            $this->kill();
-            $this->build();
+        //try {
             $this->rebuildTables($extraDataObjects);
-        }
+        // } catch (DatabaseException $ex) {
+        //     // In case of error during build force a hard reset
+        //     // e.g. pgsql doesn't allow schema updates inside transactions
+        //     $this->kill();
+        //     $this->build();
+        //     $this->rebuildTables($extraDataObjects);
+        // }
     }
 }
```

There's a couple of approaches that could be taken here.

 * Remove this try / catch, see what fails in the build matrix, and re-apply it to trigger only on that specific error
 * Change the code to kill/build only once.

Marking as a low-impact bug because it doesn't affect actual sites, only testing, and only occurs in edge-cases.

### PR DESCRIPTION
**Pull request by chillu**
Ugly, but so is the original implementation that this works around (swallowing an exception to trigger functionality)

### CODE REVIEW

**Comment by sminnee** [comment]
The change seems acceptable to me – it's no worse than the exception swallower itself.

The PGSQL failures due to a VACUUM ANALYZE - wtf the are we calling VACUUM in a test suite for!? I'd suggest attacking is issue from that angle.

**Comment by sminnee** [comment]
> wtf the are we calling VACUUM in a test suite for!? 

To answer this question: the config setting DBSchemaManager.check_and_repair_on_build is the only thing that determines whether DBSchemaManager::checkAndRepairTable() is called when updating a table during a dev/build operation, including a temp database build.

I suggest that we address this by somehow disabling check_and_repair_on_build for TempDatabases.

The approach that I would probably go with is:

 * Have an $options map argument to DBSchemaManager::schemaUpdate() where you can pass `[ 'check_and_repair_on_build' => false ]`. Note that for the implementation of that setting you will probably 
 * Have that argument set the global config flag before `executing $callback()` and reset it afterward. Note that it needs to modify the statically acessible DBSchemaManager so that `singleton(DataObjectClass)->requireTable()` is able to access it, otherwise I'd suggest a solution that doesn't involve static state manipulation.
 * When calling schemaUpdate in TempDatabase::rebuildTables(), pass `[ 'check_and_repair_on_build' => false ]`

**Comment by chillu** [comment]
Arrrghghghgh, so this was an exception that was conveniently swallowed before on Postgres, and now surfaces? Why do I always seem to pick the goddamn rabbit hole PRs :D

Thanks for digging into this Sam! Your solution isn't easy because SQLite has it's own `schemaUpdate()` implementation that would break. We could change the public API signature here, but it would require setting upper and lower constraints on framework compat with SQLite, coordinated release management, etc - not super keen on that. Instead I've just taken your approach and put it into TempDatabase directly.

**Comment by sminnee** [comment]
This gets my 👍.
